packages/web/src/components/data-entry/TextAreaV2.tsx

```diff
@@ -0,0 +1,108 @@
+import { useRef, useCallback, useEffect, ComponentPropsWithoutRef } from 'react'
+
+import cn from 'classnames'
+
+import styles from './TextAreaV2.module.css'
+
+export enum TextAreaSize {
+  MEDIUM,
+  SMALL
+}
+
+const sizeToLineHeight: Record<TextAreaSize, number> = {
+  [TextAreaSize.MEDIUM]: 24,
+  [TextAreaSize.SMALL]: 18
+}
+
+const sizeToVerticalPadding: Record<TextAreaSize, number> = {
+  [TextAreaSize.MEDIUM]: 16,
+  [TextAreaSize.SMALL]: 12
+}
+
+type TextAreaV2Props = ComponentPropsWithoutRef<'textarea'> & {
+  grows?: boolean
+  resize?: boolean
+  size?: TextAreaSize
+  heightBuffer?: number
+  maxVisibleRows?: number
+  showMaxLength?: boolean
+}
+
+export const TextAreaV2 = (props: TextAreaV2Props) => {
+  const {
+    resize = false,
+    grows = false,
+    size = TextAreaSize.MEDIUM,
+    heightBuffer,
+    maxVisibleRows,
+    maxLength,
+    showMaxLength,
+    className,
+    value,
+    children,
+    ...other
+  } = props
+
+  const ref = useRef<HTMLDivElement>(null)
+  const textareaRef = useRef<HTMLTextAreaElement>(null)
+
+  const growTextArea = useCallback(() => {
+    if (textareaRef.current) {
+      const textarea = textareaRef.current
+      textarea.style.height = '1px'
+      textarea.style.height = `${textarea.scrollHeight + (heightBuffer ?? 0)}px`
+      ref.current?.scrollTo({ top: ref.current?.scrollHeight })
+    }
+  }, [textareaRef, heightBuffer])
+
+  useEffect(() => {
+    if (grows) {
+      growTextArea()
+    }
+  }, [grows, growTextArea, value])
+
+  const style = {
+    [styles.noResize]: !resize,
+    [styles.medium]: size === TextAreaSize.MEDIUM,
+    [styles.small]: size === TextAreaSize.SMALL
+  }
+
+  const nearCharacterLimit = maxLength
+    ? `${value}`.length > (7.0 / 8.0) * maxLength
```

Comment:
nit: maybe we pull this out?

Reply:
do you mean like rip it out or make standalone and compose in via props or something?

Reply:
sorry just meant pull out the numbers into a const